sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/AlterTableAddPartitionExec.scala

```diff
@@ -0,0 +1,48 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import scala.collection.JavaConverters._
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.catalyst.analysis.PartitionAlreadyExistsException
+import org.apache.spark.sql.catalyst.expressions.Attribute
+import org.apache.spark.sql.connector.catalog.SupportsPartitions
+
+/**
+ * Physical plan node for adding partitions of table.
+ */
+case class AlterTableAddPartitionExec(
+    table: SupportsPartitions,
+    partitions: Seq[(InternalRow, Map[String, String])],
+    ignoreIfExists: Boolean) extends V2CommandExec {
+
+  override def output: Seq[Attribute] = Seq.empty
+
+  override protected def run(): Seq[InternalRow] = {
+    partitions.foreach { case (partIdent, properties) =>
+      try {
+        table.createPartition(partIdent, properties.asJava)
```

Comment:
The SQL command can add multiple partitions at once, and ideally it should be atomic.

I have 2 thoughts:
1. for v2, we don't allow the SQL command to add more than one partitions at once.
2. the v2 API should be `createPartitions` that takes an array of partitions.

Which one do you prefer? The same problem applies to drop partitions as well. @rdblue @stczwd

Reply:
FYI, the hive catalog API is `createPartitions`, which can create a list of partitions at once. Personally I prefer 2, other catalog implementation can fail with more than one partitions if they can't do it in an atomic way.

Reply:
Hm. Supporting `createPartitions` means that Table needs to support the atomic partitions operations. Once there is a problem in the middle operation, such as partition already exists,  it can roll back to the state before operations.
Hive did support atomic createPartitions, I don't know if others support this.

Reply:
I think that a variant of `createPartition` that works for multiple partitions should be added as an optional trait, or using optional methods in the existing interface.

We want sources to have predictable, standard behavior. I think that means that when adding a single partition in a group fails, the ones that were already successful should be rolled back. That way multiple calls to `createPartition` have the same result as a single call to `createPartitions` and the SQL statement has well-defined and reliable behavior.

If we agree on that behavior, then I think it is clear that Spark should handle calling either `createPartition` multiple times or calling `createPartitions` because that's the best way to get reliable behavior, while keeping table implementations simple -- those that don't support an atomic operation just implement `createPartition`.

Reply:
>  That way multiple calls to createPartition have the same result as a single call to createPartitions and the SQL statement has well-defined and reliable behavior.

It is hard to support atomic operation with multiple calls to `createPartition`, as the program may stop in the middle of operations.

How about this? We create a new optional trait `SupportsAtomicPartitions` to support multiple partition atomic operations, and for those who don't support, we should return `UnSupportOperationsException` if user trying to operate multiple partitions at the same time.

Reply:
Rolling back partition changes is like what we do with CTAS. If the write fails for non-atomic CTAS, we drop the table that was created. That won't always work, but at least the expectation is that the commands have the same behavior.

I'm okay with failing ADD PARTITION commands that have multiple partitions if atomic create/drop is not supported as well. That seems like another reasonable way to go. The important thing to me is that the commands have the same stated behavior across sources.

Reply:
> The important thing to me is that the commands have the same stated behavior across sources.

Yes, agree with that.

> We create a new optional trait SupportsAtomicPartitions to support multiple partition atomic operations, and for those who don't support, we should return UnsupportedOperationException if user trying to operate multiple partitions at the same time.

I still prefer this. @cloud-fan does this look good to you?

Reply:
`SupportsAtomicPartitions` SGTM, but we need to make the naming better.

How about `SupportsPartitionManagement` and `SupportsAtomicPartitionManagement`?

Reply:
It's OK to me. I'll change it, thanks.